"""äº”å­æ£‹æ¸¸æˆæˆ¿é—´æœåŠ¡ - åœ¨èŠå¤©æˆ¿é—´åŸºç¡€ä¸Šå¢åŠ ç®€å•å¯¹æˆ˜è§„åˆ™

è®¾è®¡ç›®æ ‡ï¼š
- æ¯ä¸ªæˆ¿é—´åªå…è®¸ä¸€åœºå¯¹æˆ˜ï¼ˆè¿›ç¨‹ç”Ÿå‘½å‘¨æœŸå†…ï¼‰ã€‚
- ä¸€å±€åªå…è®¸ä¸¤åå¯¹æˆ˜ç©å®¶ï¼šé»‘å­å’Œç™½å­ã€‚
- å…¶ä»–åŠ å…¥æˆ¿é—´çš„ç”¨æˆ·ä¸€å¾‹è§†ä¸ºè§‚æˆ˜è€…ã€‚
- ä¸€æ—¦å¯¹å±€å¼€å§‹ï¼Œä¹‹ååŠ å…¥çš„ç”¨æˆ·ä¸èƒ½å†æˆä¸ºå¯¹æˆ˜ç©å®¶ï¼Œåªèƒ½è§‚æˆ˜ã€‚

åè®®è®¾è®¡ï¼ˆåŸºäº ChatMessage è‡ªå®šä¹‰ç±»å‹ï¼Œèµ° WsEnvelope.chat é€šé“ï¼‰ï¼š
- æ™®é€šèŠå¤©ï¼šæ²¿ç”¨ USER_TEXT è¯­ä¹‰ï¼Œä¸åšé™åˆ¶ã€‚
- äº”å­æ£‹çŠ¶æ€ï¼šChatMessage.type = GOBANG_STATE_TYPEï¼Œcontent ä¸º JSONï¼š
  {
    "board": [[0,1,0,...], ...],  # 15x15 æ£‹ç›˜ï¼Œ0=ç©º,1=é»‘,2=ç™½
    "current_turn": 1/2,
    "finished": true/false,
    "winner": "black"|"white"|"",
    "role": "black"|"white"|"spectator"
  }
- äº”å­æ£‹è½å­ï¼šChatMessage.type = GOBANG_MOVE_TYPEï¼Œcontent ä¸º JSONï¼š
  {"x": 7, "y": 7}

æ³¨æ„ï¼šè¿™é‡Œæ²¡æœ‰ä¿®æ”¹ proto æ–‡ä»¶ï¼Œè€Œæ˜¯å¤ç”¨ç°æœ‰ ChatMessageï¼Œ
ä½¿ç”¨è‡ªå®šä¹‰çš„æ•´å‹ type å€¼ï¼ˆGOBANG_STATE_TYPE / GOBANG_MOVE_TYPEï¼‰ã€‚
è¿™å¯¹ protobuf æ˜¯åˆæ³•çš„ï¼ˆproto3 æ”¯æŒæœªçŸ¥æšä¸¾å€¼ï¼‰ï¼Œåç»­å¦‚æœéœ€è¦å¯ä»¥åœ¨ .proto ä¸­è¡¥å……æšä¸¾åã€‚
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass, field
from typing import Dict, Optional, Tuple, List, Set
import time
import json
import random

from fastapi import WebSocket

from protos import chat_pb2
from .chat_room import ChatRoomManager


BOARD_SIZE = 15  # æ ‡å‡† 15x15 äº”å­æ£‹
DISCONNECT_TIMEOUT_SECONDS = 300  # å¯¹æˆ˜ç©å®¶æ–­çº¿è¶…æ—¶æ—¶é—´ï¼ˆ5 åˆ†é’Ÿï¼‰

# è‡ªå®šä¹‰çš„ ChatMessage.type æ•°å€¼ï¼ˆå‰åç«¯éœ€è¦ä¿æŒä¸€è‡´ï¼‰
GOBANG_STATE_TYPE = 20
GOBANG_MOVE_TYPE = 21
GOBANG_JOIN_TYPE = 22
GOBANG_LEAVE_TYPE = 23


@dataclass
class GobangRoomState:
    """å•ä¸ªæˆ¿é—´çš„äº”å­æ£‹çŠ¶æ€"""

    # ç©å®¶èº«ä»½ï¼ˆå­˜ user_idï¼Œä¾¿äºæ–­çº¿é‡è¿åè¯†åˆ«ï¼‰
    black_user_id: Optional[int] = None
    white_user_id: Optional[int] = None

    # å·²åŠ å…¥æœ¬å±€çš„ç©å®¶ï¼ˆæœ€å¤š 2 äººï¼Œå°šæœªåˆ†é…é»‘ç™½å‰æš‚å­˜åœ¨è¿™é‡Œï¼‰
    joined_user_ids: Set[int] = field(default_factory=set)

    # å½“å‰å¯¹å±€æ˜¯å¦å·²ç»å¼€å§‹ / æ˜¯å¦å·²ç»ç»“æŸ
    started: bool = False
    finished: bool = False

    # æ£‹ç›˜ï¼š0=ç©ºï¼Œ1=é»‘ï¼Œ2=ç™½
    board: List[List[int]] = field(
        default_factory=lambda: [[0 for _ in range(BOARD_SIZE)] for _ in range(BOARD_SIZE)]
    )
    # è½®åˆ°è°ï¼š1=é»‘ï¼Œ2=ç™½
    current_turn: int = 1

    # è·èƒœæ–¹ï¼š0=æœªç»“æŸï¼Œ1=é»‘èƒœï¼Œ2=ç™½èƒœ
    winner: int = 0

    def reset_board(self) -> None:
        """é‡ç½®æ£‹ç›˜ - å½“å‰è®¾è®¡ä¸­ä¸å†è¢«è°ƒç”¨ï¼Œä»…é¢„ç•™"""
        self.board = [[0 for _ in range(BOARD_SIZE)] for _ in range(BOARD_SIZE)]
        self.current_turn = 1
        self.winner = 0


class GobangRoomManager(ChatRoomManager):
    """äº”å­æ£‹æˆ¿é—´ç®¡ç†å™¨ - ç»§æ‰¿èŠå¤©æˆ¿é—´åŠŸèƒ½ï¼Œå¢åŠ äº”å­æ£‹å¯¹æˆ˜çº¦æŸ"""

    def __init__(self) -> None:
        super().__init__()
        # room_id -> GobangRoomState
        self.room_states: Dict[int, GobangRoomState] = {}
        # æ–­çº¿è¶…æ—¶ï¼šroom_id -> (asyncio.Task, æ–­çº¿çš„ user_id)
        self._disconnect_tasks: Dict[int, Tuple[asyncio.Task, Optional[int]]] = {}

    # ---- åŸºæœ¬è¿æ¥é€»è¾‘ ----

    async def connect(self, room_id: int, websocket: WebSocket, username: str, user_id: Optional[int]) -> None:
        """è¿æ¥æˆ¿é—´ï¼šæ‰€æœ‰äººåˆå§‹éƒ½æ˜¯è§‚æˆ˜è€…ï¼Œæ˜¯å¦å‚ä¸å¯¹å±€ç”±ä¹‹åçš„â€œåŠ å…¥æ¸¸æˆâ€æ§åˆ¶"""
        # è‹¥è¯¥ç”¨æˆ·æ˜¯æ–­çº¿é‡è¿çš„å¯¹æˆ˜ç©å®¶ï¼Œå–æ¶ˆè¶…æ—¶ç»“æŸä»»åŠ¡
        self._cancel_disconnect_task_if_reconnect(room_id, user_id)

        await super().connect(room_id, websocket, username, user_id)

        state = self.room_states.setdefault(room_id, GobangRoomState())

        # åˆå§‹èº«ä»½å‡ä¸ºè§‚æˆ˜è€…ï¼Œåœ¨å‘é€ GOBANG_JOIN æ¶ˆæ¯åæ‰å†³å®šæ˜¯å¦è¿›å…¥å¯¹å±€
        await self._send_role_message(room_id, websocket, "spectator")

    def disconnect(self, room_id: int, websocket: WebSocket) -> None:
        """æ–­å¼€è¿æ¥ - è‹¥ä¸ºå¯¹æˆ˜ç©å®¶åˆ™å¯åŠ¨ 5 åˆ†é’Ÿè¶…æ—¶ï¼Œè¶…æ—¶åè‡ªåŠ¨ç»“æŸå¯¹å±€"""
        # å¿…é¡»åœ¨ super().disconnect ä¹‹å‰è·å– user_idï¼Œå¦åˆ™ä¼šè¢«æ¸…ç†
        user_id = self.websocket_to_user_id.get(websocket)
        state = self.room_states.get(room_id)

        super().disconnect(room_id, websocket)

        # è‹¥å¯¹å±€è¿›è¡Œä¸­ä¸”æ–­å¼€çš„æ˜¯é»‘/ç™½ç©å®¶ï¼Œå¯åŠ¨ 5 åˆ†é’Ÿè¶…æ—¶ä»»åŠ¡
        if (
            state
            and state.started
            and not state.finished
            and user_id is not None
            and (user_id == state.black_user_id or user_id == state.white_user_id)
        ):
            self._start_disconnect_timeout(room_id, user_id)

    # ---- æ¶ˆæ¯å¤„ç† ----

    async def handle_message(self, room_id: int, websocket: WebSocket, message: chat_pb2.ChatMessage) -> None:
        """å¤„ç†æ¶ˆæ¯ï¼šèŠå¤© / éŸ³ä¹ / äº”å­æ£‹æŒ‡ä»¤"""
        username = self.websocket_to_username.get(websocket, "Anonymous")
        user_id = self.websocket_to_user_id.get(websocket)

        # äº”å­æ£‹åŠ å…¥æ¸¸æˆ
        if message.type == GOBANG_JOIN_TYPE:
            await self._handle_gobang_join_message(room_id, websocket, username, user_id)
            return

        # äº”å­æ£‹é€€å‡ºç­‰å¾…é˜Ÿåˆ—
        if message.type == GOBANG_LEAVE_TYPE:
            await self._handle_gobang_leave_message(room_id, websocket, username, user_id)
            return

        # äº”å­æ£‹è½å­æ¶ˆæ¯
        if message.type == GOBANG_MOVE_TYPE:
            await self._handle_gobang_move_message(room_id, websocket, username, user_id, message.content)
            return

        # èŠå¤©ä¸éŸ³ä¹æ¶ˆæ¯ä»ç„¶æ­£å¸¸å¤„ç†
        if message.type in (chat_pb2.MessageType.USER_TEXT, chat_pb2.MessageType.MUSIC):
            await super().handle_message(room_id, websocket, message)
        else:
            # å…¶ä»–ç±»å‹æš‚ä¸åšæ‰©å±•ï¼Œç›´æ¥æ²¿ç”¨çˆ¶ç±»é€»è¾‘
            await super().handle_message(room_id, websocket, message)

    async def send_initial_state(self, room_id: int, websocket: WebSocket) -> None:
        """æ–°åŠ å…¥ç”¨æˆ·æ—¶ï¼Œå‘é€å½“å‰äº”å­æ£‹çŠ¶æ€ï¼ˆåŒ…å«æ£‹ç›˜ + èº«ä»½ï¼‰"""
        state = self.room_states.get(room_id)
        if not state:
            return

        user_id = self.websocket_to_user_id.get(websocket)
        payload = self._build_state_payload(room_id, state, user_id)

        state_msg = chat_pb2.ChatMessage(
            user="System",
            room_id=room_id,
            content=json.dumps(payload, ensure_ascii=False),
            timestamp=int(time.time() * 1000),
            type=GOBANG_STATE_TYPE,
        )
        await self._send_to_connection(
            room_id,
            websocket,
            chat_pb2.WsEnvelope(chat=state_msg).SerializeToString(),
        )

    # ---- å†…éƒ¨å·¥å…·æ–¹æ³• ----

    async def _broadcast_system(self, room_id: int, content: str) -> None:
        """å¹¿æ’­ç³»ç»Ÿæç¤º"""
        msg = chat_pb2.ChatMessage(
            user="System",
            room_id=room_id,
            content=content,
            timestamp=int(time.time() * 1000),
            type=chat_pb2.MessageType.SYSTEM,
        )
        await self.broadcast(room_id, chat_pb2.WsEnvelope(chat=msg).SerializeToString())

    async def _send_role_message(self, room_id: int, websocket: WebSocket, role: str) -> None:
        """å•ç‹¬å‘æŸä¸ªç”¨æˆ·è¯´æ˜å…¶åœ¨æœ¬æˆ¿é—´ä¸­çš„èº«ä»½"""
        role_desc = {
            "black": "ä½ æ˜¯æœ¬å±€äº”å­æ£‹çš„é»‘å­ç©å®¶ï¼ˆå…ˆæ‰‹ï¼‰",
            "white": "ä½ æ˜¯æœ¬å±€äº”å­æ£‹çš„ç™½å­ç©å®¶ï¼ˆåæ‰‹ï¼‰",
            "spectator": "ä½ æ˜¯è§‚æˆ˜è€…ï¼Œæœ¬å±€è½å­æƒåªå±äºé»‘å­å’Œç™½å­ç©å®¶",
        }.get(role, "ä½ æ˜¯è§‚æˆ˜è€…ï¼Œæœ¬å±€è½å­æƒåªå±äºé»‘å­å’Œç™½å­ç©å®¶")

        msg = chat_pb2.ChatMessage(
            user="System",
            room_id=room_id,
            content=role_desc,
            timestamp=int(time.time() * 1000),
            type=chat_pb2.MessageType.SYSTEM,
        )
        await self._send_to_connection(room_id, websocket, chat_pb2.WsEnvelope(chat=msg).SerializeToString())

    async def _handle_gobang_move_message(
        self,
        room_id: int,
        websocket: WebSocket,
        username: str,
        user_id: Optional[int],
        content: str,
    ) -> None:
        """è§£æå¹¶å¤„ç†äº”å­æ£‹è½å­æŒ‡ä»¤ï¼ˆcontent ä¸º JSON: {"x": int, "y": int}ï¼‰"""
        state = self.room_states.setdefault(room_id, GobangRoomState())

        # ä¸€å±€ç»“æŸåä¸å†æ¥å—ä»»ä½•è½å­ï¼ˆä¿æŒâ€œæ¯æˆ¿é—´ä¸€å±€â€çš„çº¦æŸï¼‰
        if state.finished:
            await self._send_error(room_id, websocket, "æœ¬æˆ¿é—´çš„äº”å­æ£‹å¯¹å±€å·²ç»ç»“æŸï¼Œä¸èƒ½å†è½å­ã€‚")
            return

        # å¯¹å±€å°šæœªå¼€å§‹æˆ–ç©å®¶èº«ä»½ä¸è¶³
        if not state.started or state.black_user_id is None or state.white_user_id is None:
            await self._send_error(room_id, websocket, "äº”å­æ£‹å¯¹å±€å°šæœªå¼€å§‹æˆ–ç©å®¶å°šæœªå°±ä½ã€‚")
            return

        if user_id is None:
            await self._send_error(room_id, websocket, "æœªç™»å½•ç”¨æˆ·ä¸èƒ½å‚ä¸äº”å­æ£‹å¯¹å±€ï¼Œåªèƒ½è§‚æˆ˜ã€‚")
            return

        # åˆ¤æ–­å½“å‰ç”¨æˆ·æ˜¯å¦ä¸ºé»‘/ç™½ç©å®¶
        if user_id == state.black_user_id:
            player_color = 1
        elif user_id == state.white_user_id:
            player_color = 2
        else:
            await self._send_error(room_id, websocket, "ä½ ä¸æ˜¯æœ¬å±€çš„å¯¹æˆ˜ç©å®¶ï¼Œåªèƒ½è§‚æˆ˜ã€‚")
            return

        # è½®åˆ°è°è½å­
        if player_color != state.current_turn:
            await self._send_error(room_id, websocket, "è¿˜æ²¡è½®åˆ°ä½ è½å­ã€‚")
            return

        # è§£æåæ ‡
        try:
            data = json.loads(content or "{}")
            x = int(data.get("x"))
            y = int(data.get("y"))
        except Exception:
            await self._send_error(room_id, websocket, "æŒ‡ä»¤æ ¼å¼é”™è¯¯ï¼Œåº”ä¸º JSON: {\"x\": 7, \"y\": 7}")
            return

        if not (0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE):
            await self._send_error(room_id, websocket, f"åæ ‡è¶Šç•Œï¼Œåˆæ³•èŒƒå›´ä¸º [0, {BOARD_SIZE - 1}]ã€‚")
            return

        if state.board[y][x] != 0:
            await self._send_error(room_id, websocket, "è¯¥ä½ç½®å·²ç»æœ‰æ£‹å­äº†ï¼Œè¯·é€‰æ‹©å…¶ä»–ä½ç½®ã€‚")
            return

        # è½å­
        state.board[y][x] = player_color

        # æ£€æŸ¥æ˜¯å¦äº”è¿
        if self._check_winner(state.board, x, y, player_color):
            state.finished = True
            state.winner = player_color

            # è·å–åŒæ–¹ç”¨æˆ·åï¼ˆåœ¨é‡ç½®å‰ï¼‰
            black_name = self._get_username_by_user_id(room_id, state.black_user_id)
            white_name = self._get_username_by_user_id(room_id, state.white_user_id)
            winner_name = black_name if player_color == 1 else white_name

            game_over_msg = (
                f"ğŸ® å¯¹å±€ç»“æŸï¼é»‘æ–¹ï¼š{black_name} vs ç™½æ–¹ï¼š{white_name} â€”â€” "
                f"{winner_name}ï¼ˆ{'é»‘å­' if player_color == 1 else 'ç™½å­'}ï¼‰è·èƒœï¼å¯ç‚¹å‡»ã€ŒåŠ å…¥å¯¹å±€ã€å¼€å§‹æ–°ä¸€å±€ã€‚"
            )
            await self._broadcast_system(room_id, game_over_msg)

            # åŒæ—¶ä½œä¸ºèŠå¤©æ¶ˆæ¯å¹¿æ’­ï¼Œä¾¿äºåœ¨æ¶ˆæ¯åˆ—è¡¨ä¸­æŸ¥çœ‹
            chat_msg = chat_pb2.ChatMessage(
                user="System",
                room_id=room_id,
                content=game_over_msg,
                timestamp=int(time.time() * 1000),
                type=chat_pb2.MessageType.USER_TEXT,
            )
            await self.broadcast(room_id, chat_pb2.WsEnvelope(chat=chat_msg).SerializeToString())

            # é‡ç½®çŠ¶æ€ï¼Œå…è®¸é‡æ–°åŠ å…¥å¯¹å±€
            state.black_user_id = None
            state.white_user_id = None
            state.joined_user_ids.clear()
            state.reset_board()
            state.started = False
            state.finished = False
            state.winner = 0
        else:
            # è½®åˆ°å¦ä¸€æ–¹
            state.current_turn = 2 if state.current_turn == 1 else 1
            next_desc = "é»‘å­" if state.current_turn == 1 else "ç™½å­"
            await self._broadcast_system(
                room_id,
                f"{username} åœ¨ ({x}, {y}) è½å­æˆåŠŸï¼Œä¸‹ä¸€æ‰‹è½®åˆ° {next_desc}ã€‚",
            )

        # å‘æˆ¿é—´å†…æ‰€æœ‰ç”¨æˆ·å¹¿æ’­æœ€æ–°æ£‹ç›˜çŠ¶æ€ï¼ˆåŒ…å«å„è‡ªçš„ roleï¼‰
        await self._broadcast_gobang_state(room_id)

    async def _handle_gobang_join_message(
        self,
        room_id: int,
        websocket: WebSocket,
        username: str,
        user_id: Optional[int],
    ) -> None:
        """å¤„ç†åŠ å…¥äº”å­æ£‹å¯¹å±€çš„è¯·æ±‚"""
        state = self.room_states.setdefault(room_id, GobangRoomState())

        if user_id is None:
            await self._send_error(room_id, websocket, "æœªç™»å½•ç”¨æˆ·ä¸èƒ½åŠ å…¥å¯¹å±€ï¼Œåªèƒ½è§‚æˆ˜ã€‚")
            return

        if state.finished:
            await self._send_error(room_id, websocket, "æœ¬å±€å·²ç»ç»“æŸï¼Œä¸èƒ½å†åŠ å…¥ï¼Œåªèƒ½è§‚æˆ˜ã€‚")
            return

        # å·²ç»åœ¨æœ¬å±€ä¸­ï¼ˆé»‘/ç™½/å·²åŠ å…¥ç­‰å¾…å¼€å±€ï¼‰ï¼Œç›´æ¥è¿”å›å½“å‰çŠ¶æ€å³å¯
        if user_id == state.black_user_id or user_id == state.white_user_id or user_id in state.joined_user_ids:
            await self._send_error(room_id, websocket, "ä½ å·²ç»åœ¨æœ¬å±€ä¸­ï¼Œæ— éœ€é‡å¤åŠ å…¥ã€‚")
            return

        # å¦‚æœå·²ç»æœ‰ä¸¤ä¸ªç©å®¶åœ¨å¯¹å±€ä¸­ï¼ˆé»‘ç™½å·²ç¡®å®šï¼‰ï¼Œä¸å†å…è®¸ç¬¬ä¸‰äººåŠ å…¥ï¼Œåªèƒ½è§‚æˆ˜
        if state.black_user_id is not None and state.white_user_id is not None:
            await self._send_error(room_id, websocket, "æœ¬å±€å·²æ»¡ä¸¤åç©å®¶ï¼Œä½ åªèƒ½ä½œä¸ºè§‚æˆ˜è€…ã€‚")
            return

        # è®°å½•ä¸ºå·²ç”³è¯·åŠ å…¥çš„ç©å®¶
        state.joined_user_ids.add(user_id)

        # å¦‚æœå½“å‰åŠ å…¥äººæ•°ä¸è¶³ 2ï¼Œç­‰å¾…å¦ä¸€äºº
        if len(state.joined_user_ids) < 2:
            await self._send_error(room_id, websocket, "å·²åŠ å…¥å¯¹å±€ï¼Œç­‰å¾…å¦ä¸€ä½ç©å®¶åŠ å…¥...")
            # åŒæ—¶å¹¿æ’­ä¸€æ¡ç³»ç»Ÿæç¤º
            await self._broadcast_system(room_id, f"{username} å·²åŠ å…¥æœ¬å±€ï¼Œç­‰å¾…å¦ä¸€ä½ç©å®¶...")
            # å•ç‹¬å‘è¯¥ç©å®¶åŒæ­¥å½“å‰çŠ¶æ€ï¼ˆä»æ˜¯è§‚æˆ˜è€…èº«ä»½ï¼‰
            await self._broadcast_gobang_state(room_id)
            return

        # æ°å¥½ä¸¤äººï¼Œéšæœºåˆ†é…é»‘ç™½å¹¶å¼€å§‹å¯¹å±€
        players = list(state.joined_user_ids)[:2]
        random.shuffle(players)
        state.black_user_id, state.white_user_id = players[0], players[1]
        state.started = True
        state.current_turn = 1

        await self._broadcast_system(
            room_id,
            f"äº”å­æ£‹å¯¹å±€å¼€å§‹ï¼šé»‘å­ï¼ˆuser_id={state.black_user_id}ï¼‰ï¼Œç™½å­ï¼ˆuser_id={state.white_user_id}ï¼‰ã€‚",
        )

        # å¹¿æ’­æœ€æ–°çŠ¶æ€ç»™æ‰€æœ‰äººï¼ˆåŒ…æ‹¬è§‚æˆ˜è€…ï¼‰
        await self._broadcast_gobang_state(room_id)

    async def _handle_gobang_leave_message(
        self,
        room_id: int,
        websocket: WebSocket,
        username: str,
        user_id: Optional[int],
    ) -> None:
        """å¤„ç†é€€å‡ºäº”å­æ£‹ç­‰å¾…é˜Ÿåˆ—çš„è¯·æ±‚ï¼ˆä»…åœ¨å¯¹å±€æœªå¼€å§‹æ—¶å…è®¸ï¼‰"""
        state = self.room_states.setdefault(room_id, GobangRoomState())

        if user_id is None:
            await self._send_error(room_id, websocket, "æœªç™»å½•ç”¨æˆ·ä¸èƒ½é€€å‡ºå¯¹å±€ã€‚")
            return

        # å¯¹å±€å·²ç»å¼€å§‹åˆ™ä¸å…è®¸é€€å‡ºï¼ˆé˜²æ­¢ä¸­é€”è§£æ•£ï¼‰
        if state.started:
            await self._send_error(room_id, websocket, "å¯¹å±€å·²å¼€å§‹ï¼Œä¸èƒ½é€€å‡ºå¯¹å±€ã€‚")
            return

        if user_id not in state.joined_user_ids:
            await self._send_error(room_id, websocket, "ä½ å½“å‰æœªåœ¨å¯¹å±€ç­‰å¾…é˜Ÿåˆ—ä¸­ï¼Œæ— éœ€é€€å‡ºã€‚")
            return

        # ä»ç­‰å¾…é˜Ÿåˆ—ä¸­ç§»é™¤
        state.joined_user_ids.discard(user_id)

        await self._broadcast_system(room_id, f"{username} é€€å‡ºäº†æœ¬å±€ç­‰å¾…é˜Ÿåˆ—ã€‚")
        await self._broadcast_gobang_state(room_id)

    def _cancel_disconnect_task_if_reconnect(self, room_id: int, user_id: Optional[int]) -> None:
        """è‹¥é‡è¿ç”¨æˆ·æ­£æ˜¯æ–­çº¿è¶…æ—¶ç­‰å¾…çš„ç©å®¶ï¼Œå–æ¶ˆè¶…æ—¶ä»»åŠ¡"""
        if user_id is None:
            return
        entry = self._disconnect_tasks.pop(room_id, None)
        if entry:
            task, disconnected_uid = entry
            if disconnected_uid == user_id and not task.done():
                task.cancel()

    def _start_disconnect_timeout(self, room_id: int, disconnected_user_id: int) -> None:
        """å¯åŠ¨æ–­çº¿è¶…æ—¶ä»»åŠ¡ï¼š5 åˆ†é’Ÿåè‹¥æœªé‡è¿åˆ™è‡ªåŠ¨ç»“æŸå¯¹å±€"""
        # è‹¥å·²æœ‰è¶…æ—¶ä»»åŠ¡ï¼Œå…ˆå–æ¶ˆ
        entry = self._disconnect_tasks.pop(room_id, None)
        if entry:
            task, _ = entry
            if not task.done():
                task.cancel()

        async def _timeout_task() -> None:
            try:
                await asyncio.sleep(DISCONNECT_TIMEOUT_SECONDS)
                await self._end_game_due_to_disconnect(room_id, disconnected_user_id)
            except asyncio.CancelledError:
                pass
            finally:
                self._disconnect_tasks.pop(room_id, None)

        task = asyncio.create_task(_timeout_task())
        self._disconnect_tasks[room_id] = (task, disconnected_user_id)

    async def _end_game_due_to_disconnect(self, room_id: int, disconnected_user_id: int) -> None:
        """å› å¯¹æˆ˜ç©å®¶æ–­çº¿è¶…æ—¶è€Œç»“æŸå¯¹å±€ï¼Œé‡ç½®çŠ¶æ€å¹¶å¹¿æ’­"""
        state = self.room_states.get(room_id)
        if not state or not state.started or state.finished:
            return

        disconnected_name = self._get_username_by_user_id(room_id, disconnected_user_id)
        other_user_id = (
            state.white_user_id if disconnected_user_id == state.black_user_id else state.black_user_id
        )
        other_name = self._get_username_by_user_id(room_id, other_user_id)
        role_desc = "é»‘æ–¹" if disconnected_user_id == state.black_user_id else "ç™½æ–¹"

        game_over_msg = (
            f"â± å¯¹å±€ç»“æŸï¼{disconnected_name}ï¼ˆ{role_desc}ï¼‰æ–­çº¿è¶…è¿‡ 5 åˆ†é’Ÿï¼Œ"
            f"å¦ä¸€æ–¹ {other_name} è·èƒœã€‚å¯ç‚¹å‡»ã€ŒåŠ å…¥å¯¹å±€ã€å¼€å§‹æ–°ä¸€å±€ã€‚"
        )
        await self._broadcast_system(room_id, game_over_msg)

        chat_msg = chat_pb2.ChatMessage(
            user="System",
            room_id=room_id,
            content=game_over_msg,
            timestamp=int(time.time() * 1000),
            type=chat_pb2.MessageType.USER_TEXT,
        )
        await self.broadcast(room_id, chat_pb2.WsEnvelope(chat=chat_msg).SerializeToString())

        # é‡ç½®çŠ¶æ€
        state.black_user_id = None
        state.white_user_id = None
        state.joined_user_ids.clear()
        state.reset_board()
        state.started = False
        state.finished = False
        state.winner = 0

        await self._broadcast_gobang_state(room_id)

    def _get_username_by_user_id(self, room_id: int, user_id: Optional[int]) -> str:
        """æ ¹æ® user_id è·å–å½“å‰åœ¨æˆ¿é—´å†…çš„ç”¨æˆ·åï¼Œè‹¥ä¸åœ¨çº¿åˆ™è¿”å›å ä½"""
        if user_id is None:
            return "æœªçŸ¥"
        for ws in self.room_id_to_connections.get(room_id, set()):
            if self.websocket_to_user_id.get(ws) == user_id:
                return self.websocket_to_username.get(ws, "æœªçŸ¥")
        return f"ç”¨æˆ·{user_id}"

    async def _send_error(self, room_id: int, websocket: WebSocket, message: str) -> None:
        """å‘é€é”™è¯¯æç¤ºï¼ˆç³»ç»Ÿæ¶ˆæ¯ï¼Œä»…å‘ç»™å½“å‰ç”¨æˆ·ï¼‰"""
        msg = chat_pb2.ChatMessage(
            user="System",
            room_id=room_id,
            content=message,
            timestamp=int(time.time() * 1000),
            type=chat_pb2.MessageType.SYSTEM,
        )
        await self._send_to_connection(room_id, websocket, chat_pb2.WsEnvelope(chat=msg).SerializeToString())

    def _build_state_payload(
        self,
        room_id: int,
        state: GobangRoomState,
        user_id: Optional[int],
    ) -> Dict:
        """æ„é€ å‘é€ç»™æŸä¸ªç”¨æˆ·çš„äº”å­æ£‹çŠ¶æ€ JSON payload"""
        if user_id is not None:
            if user_id == state.black_user_id:
                role = "black"
            elif user_id == state.white_user_id:
                role = "white"
            elif user_id in state.joined_user_ids:
                role = "waiting_player"  # å·²åŠ å…¥ç­‰å¾…é˜Ÿåˆ—ï¼Œæ˜¾ç¤ºä¸ºç©å®¶
            else:
                role = "spectator"
        else:
            role = "spectator"

        if state.winner == 1:
            winner_str = "black"
        elif state.winner == 2:
            winner_str = "white"
        else:
            winner_str = ""

        return {
            "board": state.board,
            "current_turn": state.current_turn,
            "finished": state.finished,
            "winner": winner_str,
            "role": role,
            "room_id": room_id,
            "started": state.started,
        }

    async def _broadcast_gobang_state(self, room_id: int) -> None:
        """æ ¹æ®å½“å‰çŠ¶æ€å‘æˆ¿é—´å†…æ‰€æœ‰è¿æ¥å¹¿æ’­äº”å­æ£‹æ£‹ç›˜çŠ¶æ€"""
        state = self.room_states.get(room_id)
        if not state:
          return

        connections = list(self.room_id_to_connections.get(room_id, set()))
        if not connections:
            return

        for ws in connections:
            uid = self.websocket_to_user_id.get(ws)
            payload = self._build_state_payload(room_id, state, uid)
            msg = chat_pb2.ChatMessage(
                user="System",
                room_id=room_id,
                content=json.dumps(payload, ensure_ascii=False),
                timestamp=int(time.time() * 1000),
                type=GOBANG_STATE_TYPE,
            )
            try:
                await ws.send_bytes(chat_pb2.WsEnvelope(chat=msg).SerializeToString())
            except Exception:
                # å‘é€å¤±è´¥æ—¶äº¤ç»™åŸºç¡€ ChatRoomManager æ¸…ç†è¿æ¥
                self.disconnect(room_id, ws)

    # ---- äº”å­æ£‹è§„åˆ™æ ¡éªŒ ----

    def _check_winner(self, board: List[List[int]], x: int, y: int, color: int) -> bool:
        """åˆ¤æ–­åœ¨ (x, y) ä½ç½®è½ä¸‹ color åæ˜¯å¦å·²ç»äº”è¿"""
        directions: List[Tuple[int, int]] = [
            (1, 0),   # æ°´å¹³
            (0, 1),   # å‚ç›´
            (1, 1),   # æ­£æ–œçº¿
            (1, -1),  # åæ–œçº¿
        ]
        for dx, dy in directions:
            count = 1
            # æ­£å‘
            count += self._count_dir(board, x, y, dx, dy, color)
            # åå‘
            count += self._count_dir(board, x, y, -dx, -dy, color)
            if count >= 5:
                return True
        return False

    def _count_dir(
        self,
        board: List[List[int]],
        x: int,
        y: int,
        dx: int,
        dy: int,
        color: int,
    ) -> int:
        """æ²¿æŸä¸ªæ–¹å‘ç»Ÿè®¡è¿ç»­ç›¸åŒé¢œè‰²çš„æ£‹å­æ•°é‡ï¼ˆä¸å«èµ·ç‚¹ï¼‰"""
        cnt = 0
        cx, cy = x + dx, y + dy
        while 0 <= cx < BOARD_SIZE and 0 <= cy < BOARD_SIZE and board[cy][cx] == color:
            cnt += 1
            cx += dx
            cy += dy
        return cnt


# å…¨å±€å®ä¾‹
gobang_room_manager = GobangRoomManager()

